packages/@react-stately/table/src/useTreeGridState.ts

```diff
@@ -10,73 +10,91 @@
-import {useCollection} from '@react-stately/collections';
 import {useControlledState} from '@react-stately/utils';
 
 export interface TreeGridState<T> extends TableState<T> {
   /** A set of keys for items that are expanded. */
   expandedKeys: 'all' | Set<Key>,
   /** Toggles the expanded state for a row by its key. */
   toggleKey(key: Key): void,
-  collection: ITreeGridCollection<T>
+  /** The key map containing nodes representing the collection's tree grid structure. */
+  keyMap: Map<Key, GridNode<T>>,
+  /** The number of columns provided by the user. */
+  userColumnCount: number
 }
 
-export interface TreeGridStateProps<T> extends Expandable, TableStateProps<T> {
-  collection?: ITreeGridCollection<T>
-}
+// TODO: For now I am omitting collection since this is a bit of a special case where are still using TableCollection but include
+// extra information for the tree grid in the state. That means the user can't really provide their own collection. Perhaps it would
+// be good to still have a TreeGridCollection implementation in that case?
+export interface TreeGridStateProps<T> extends Expandable, Omit<TableStateProps<T>, 'collection'> {}
 
 /**
  * Provides state management for a tree grid component. Handles building a collection
  * of columns and rows from props. In addition, it tracks and manages expanded rows, row selection, and sort order changes.
  */
 export function useTreeGridState<T extends object>(props: TreeGridStateProps<T>): TreeGridState<T> {
+  let {
+    selectionMode = 'none',
+    showSelectionCheckboxes,
+    showDragButtons,
+    expandedKeys: propExpandedKeys,
+    defaultExpandedKeys: propDefaultExpandedKeys,
+    onExpandedChange,
+    children
+  } = props;
+
   let [expandedKeys, setExpandedKeys] = useControlledState(
-    props.expandedKeys ? convertExpanded(props.expandedKeys) : undefined,
-    props.defaultExpandedKeys ? convertExpanded(props.defaultExpandedKeys) : new Set(),
-    props.onExpandedChange
+    propExpandedKeys ? convertExpanded(propExpandedKeys) : undefined,
+    propDefaultExpandedKeys ? convertExpanded(propDefaultExpandedKeys) : new Set(),
+    onExpandedChange
   );
-  let {selectionMode = 'none', showSelectionCheckboxes, showDragButtons} = props;
 
   let context = useMemo(() => ({
     showSelectionCheckboxes: showSelectionCheckboxes && selectionMode !== 'none',
     showDragButtons: showDragButtons,
     selectionMode,
     columns: []
   // eslint-disable-next-line react-hooks/exhaustive-deps
-  }), [props.children, showSelectionCheckboxes, selectionMode, showDragButtons]);
+  }), [children, showSelectionCheckboxes, selectionMode, showDragButtons]);
 
-  let collection = useCollection<T, ITreeGridCollection<T>>(
-    props,
-    useCallback((nodes) => new TreeGridCollection(nodes, {...context, expandedKeys}), [context, expandedKeys]),
-    context
-  );
+  let builder = useMemo(() => new CollectionBuilder<T>(), []);
```

Comment:
Since we don't have a TreeGridCollection anymore, we need to call CollectionBuilder to generate the table's nodes so we can generate the treegrid keymap. Note that this means I have to export CollectionBuilder from `react-stately/collections` which wasn't being exported before